File: backend/app/models/schema/billing.py
from datetime import datetime
from typing import Optional, Dict, List
from beanie import Document, Indexed
from pydantic import BaseModel, Field


class ModelPricingCreate(BaseModel):
    model_name: str
    prompt_token_cost: float
    completion_token_cost: float
    description: Optional[str] = None


class ModelPricingUpdate(BaseModel):
    prompt_token_cost: Optional[float] = None
    completion_token_cost: Optional[float] = None
    description: Optional[str] = None


class ModelPricingStatusUpdate(BaseModel):
    status: str


class TokenUsage(BaseModel):
    prompt_tokens: int = 0
    completion_tokens: int = 0
    total_tokens: int = 0


class CostBreakdown(BaseModel):
    prompt_cost: float = 0.0
    completion_cost: float = 0.0
    total_cost: float = 0.0


class ModelUsage(BaseModel):
    requests: int = 0
    tokens: int = 0
    cost: float = 0.0


class TimeSeriesEntry(BaseModel):
    timestamp: datetime
    total_tokens: TokenUsage
    cost: CostBreakdown
    models_used: Dict[str, ModelUsage]


class DailyUsage(BaseModel):
    date: str
    total_tokens: TokenUsage
    cost: CostBreakdown
    models_used: Dict[str, ModelUsage]


class MonthlyUsage(BaseModel):
    month: str
    total_tokens: TokenUsage
    cost: CostBreakdown
    models_used: Dict[str, ModelUsage]


class OverallStats(BaseModel):
    total_tokens: int
    total_cost: float
    average_tokens_per_request: float
    total_requests: int


class TimeSeriesData(BaseModel):
    hourly: List[TimeSeriesEntry]
    daily: List[DailyUsage]
    monthly: List[MonthlyUsage]


class UsageAnalytics(BaseModel):
    overall_stats: OverallStats
    time_series_data: TimeSeriesData

File: frontend/src/lib/types/billing.ts
export interface TokenMetrics {
  prompt_tokens: number;
  completion_tokens: number;
  total_tokens: number;
}

export interface CostMetrics {
  prompt_cost: number;
  completion_cost: number;
  total_cost: number;
}

export interface ModelUsage {
  requests: number;
  tokens: number;
  cost: number;
}

export interface TimeSeriesEntry {
  timestamp?: string;
  date?: string;
  month?: string;
  total_tokens: TokenMetrics;
  cost: CostMetrics;
  models_used: Record<string, ModelUsage>;
}

export interface BillingData {
  overall_stats: {
    total_tokens: number;
    total_cost: number;
    average_tokens_per_request: number;
    total_requests: number;
  };
  time_series_data: {
    hourly: TimeSeriesEntry[];
    daily: TimeSeriesEntry[];
    monthly: TimeSeriesEntry[];
  };
}
